feat(forms): change addControl & removeControl to accept options

MikeJerred opened the pull request:
This commit adds an optional argument to the methods addControl
and removeControl in the FormGroup class. This argument can be
used to prevent an event from being emitted.

PR Close #29662

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #29662 


## What is the new behavior?
An argument can now be specified to prevent an event from being emitted.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

MikeJerred committed:
feat(forms): change addControl & removeControl to accept options

This commit adds an optional argument to the methods addControl
and removeControl in the FormGroup class. This argument can be
used to prevent an event from being emitted.

PR Close #29662

MikeJerred committed:
feat(forms): change addControl & removeControl to accept options

This commit fixes a linting error.

PR Close #29662

MikeJerred committed:
feat(forms): change addControl & removeControl to accept options

This commit fixes an error from public_api_guard

PR Close #29662

kara commented on packages/forms/src/model.ts:
```diff
@@ -1281,9 +1281,19 @@ export class FormGroup extends AbstractControl {
+   *
+   * * `emitEvent`: When true or not supplied (the default), both the `statusChanges` and
+   * `valueChanges`
+   * observables emit events with the latest status and value when the control is added.
```
Nit: formatting

kara commented on packages/forms/src/model.ts:
```diff
@@ -1281,9 +1281,19 @@ export class FormGroup extends AbstractControl {
+  addControl(name: string, control: AbstractControl, options: {emitEvent?: boolean} = {}): void {
     this.registerControl(name, control);
+
+    if (options.emitEvent === false) return;
```
If we simply return early here, the value and validity of the `FormGroup` will not be updated. This seems like unwanted behavior to me because it breaks our guarantee that the value and validity state of a `FormGroup` is built up based on its child controls (it will be incorrect). It also changes the behavior of the `emitEvent` flag to be inconsistent with the same flag for other methods, which only control events.

I think what you want to do is pass the options to the `updateValueAndValidity` call, so the value/validity are updated, but the events are skipped (only). Same below.

aitboudad replied:
still useful in some use-cases, when dealing with a large form calling updateValueAndValidity takes a quick large amount of time, so instead we call `updateValueAndValidity` once we have done with registration.

so having an option to skip `updateValueAndValidity` would be great!

kara replied:
Seems like that would be a larger API design discussion, and outside the scope of this PR.

MikeJerred replied:
It sounds like there should be an options parameter added to `addControl` and `removeControl`, and also there should be the `registerControl` and `unregisterControl` methods?

MikeJerred replied:
Hi, just want to check up on this: do the changes I have made here seem correct now? If so I can do them for FormArray too.

kara commented on packages/forms/test/form_group_spec.ts:
```diff
@@ -170,6 +170,32 @@ import {of } from 'rxjs';
+        g.addControl('two', new FormControl('2', Validators.minLength(10)), {emitEvent: false});
+
+        expect(g.value).toEqual({'one': '1'});
+        expect(g.valid).toBe(true);
```
These tests don't seem to be quite right. I'd argue that the value and validity *should* be updated, but the events should not be emitted. Let's change this test to assert the opposite and add a new test that actually tests event emission?

kara reviewed (changes requested):
Thanks for doing this! I have some thoughts below. Main things are:

- `emitEvent` flag should only control event emission, not value/validity updates
- We should probably add a similar feature to `FormArray` so they are consistent.

MikeJerred committed:
fix(forms): fix addControl & removeControl

Fix the behavior of the addControl and removeControl methods in FormGroup
so that they pass the opts argument to updateValueAndValidity.
Also add an unregisterControl method.

PR Close #29662

MikeJerred committed:
fix(forms): fix error in public_api_guard

Update the public_api_guard FormGroup methods for
addControl, removeControl and unregisterControl.

PR Close #29662

MikeJerred commented:
I have made changes so that addControl/removeControl pass the options argument to updateValueAndValidity, and also added an unregister method which does not update value and validity. If this approach seems sensible I will apply them to FormArray too.

kara commented on packages/forms/src/model.ts:
```diff
@@ -1274,29 +1274,64 @@ export class FormGroup extends AbstractControl {
+   *
+   * @param name The control name to unregister from the collection
+   */
+  unregisterControl(name: string): AbstractControl|null {
```
Can you split this out into a separate PR? We try to make atomic changes, and this is a new API that needs discussion.

kara commented on packages/forms/src/model.ts:
```diff
@@ -1274,29 +1274,64 @@ export class FormGroup extends AbstractControl {
-  addControl(name: string, control: AbstractControl): void {
+  addControl(
+      name: string, control: AbstractControl,
+      opts: {onlySelf?: boolean, emitEvent?: boolean} = {}): void {
```
Can you split out `onlySelf` into a separate change? `emitEvent` was the original scope of the PR, and the other option needs further discussion.

kara reviewed (changes requested):
The `emitEvent` option looks good, but let's remove the other new APIs added since the last review into other PRs / issues.

MikeJerred committed:
fix(forms): remove out of scope apis

PR Close #29662

MikeJerred committed:
feat(forms): add emitEvent option to FormArray

This commit adds the emitEvent option to all methods that add/remove controls
from FormArray.

PR Close #29662

MikeJerred committed:
fix(forms): fix unit tests

This commit fixes the form array unit tests to use fakeAsync correctly.

PR Close #29662

MikeJerred committed:
fix(forms): format code

This comment fixes formatting errors.

PR Close #29662

MikeJerred committed:
fix(forms): format code

This comment fixes formatting errors.

PR Close #29662

MikeJerred committed:
fix(forms): format code

This comment fixes formatting errors.

PR Close #29662

MikeJerred commented:
@kara I have made these changes, and extended them to FormArray too

kara reviewed (approved):
Changes LGTM. Can you squash the commits such that there are the following commits:

1 - feat(forms): add emitEvent option when adding or removing controls from a group
2 - feat(forms): add emitEvent option when adding or removing controls from an array

We probably also want to add some more detail to the commit messages (e.g. why this feature is useful).

kara commented:
@IgorMinar Do you want to take a look at the public API changes?

MikeJerred commented:
@kara I have squashed the commits but github has closed the PR when I did this as I deleted the branch on remote

MikeJerred commented:
@kara I have created a new PR here: https://github.com/angular/angular/pull/30623
